zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1098 @@
+
+/// A checkpointing block verifier.
+///
+/// Verifies blocks using a supplied list of checkpoints. There must be at
+/// least one checkpoint for the genesis block.
+#[derive(Debug)]
+struct CheckpointVerifier {
+    // Inputs
+    //
     /// Each checkpoint consists of a coinbase height and block header hash.
     ///
     /// Checkpoints should be chosen to avoid forks or chain reorganizations,
     /// which only happen in the last few hundred blocks in the chain.
     /// (zcashd allows chain reorganizations up to 99 blocks, and prunes
     /// orphaned side-chains after 288 blocks.)
-    checkpoint_list: Arc<HashMap<BlockHeight, BlockHeaderHash>>,
+    ///
+    /// There must be a checkpoint for the genesis block at BlockHeight 0.
+    /// (All other checkpoints are optional.)
+    checkpoint_list: BTreeMap<BlockHeight, BlockHeaderHash>,
+
+    // Queued Blocks
+    //
+    /// A queue of unverified blocks.
+    ///
+    /// Contains a list of unverified blocks at each block height. In most cases,
+    /// the checkpoint verifier will store zero or one block at each height.
+    ///
+    /// Blocks are verified in order, when there is a chain from the next
+    /// checkpoint, back to the `current_checkpoint_height`.Each pair of
+    /// checkpoints is used to verify all the blocks between those checkpoints.
+    ///
+    /// The first checkpoint does not have any ancestors, so it only verifies the
+    /// genesis block.
+    queued: BTreeMap<BlockHeight, QueuedBlockList>,
+
+    /// The range of heights that we are currently verifying. Extends from the
+    /// most recently verified checkpoint (`Excluded`), to the next highest
+    /// checkpoint (`Included`).
+    ///
+    /// If checkpoint verification has not started yet, the current range only
+    /// contains the genesis checkpoint.
+    ///
+    /// `None` means that checkpoint verification has finished.
+    current_checkpoint_range: Option<VerifyBounds>,
 }
 
-/// The error type for the CheckpointVerifier Service.
-// TODO(jlusby): Error = Report ?
-type Error = Box<dyn error::Error + Send + Sync + 'static>;
+/// The CheckpointVerifier implementation.
+///
+/// Contains non-service utility functions for CheckpointVerifiers.
+impl CheckpointVerifier {
+    /// Return a checkpoint verification service, using the provided `checkpoint_list`.
+    ///
+    /// The returned type is opaque to allow instrumentation or other wrappers, but
+    /// can be boxed for storage. It is also `Clone` to allow sharing of a
+    /// verification service.
+    ///
+    /// This function should be called only once for a particular checkpoint list (and
+    /// network), rather than constructing multiple verification services based on the
+    /// same checkpoint list.
+    //
+    // Currently only used in tests.
+    //
+    // We'll use this function in the overall verifier, which will split blocks
+    // between BlockVerifier and CheckpointVerifier.
+    #[cfg(test)]
```

Comment:
I think it'd be preferable not to do cfg-gating here.

Reply:
I'll remove `cfg(test)` when we integrate checkpointing with block verification:
https://github.com/ZcashFoundation/zebra/pull/524#discussion_r448096570